File: DesignPatterns/5-templateMethodPattern/python/template.py
#!/usr/bin/env python3
# -*- coding:utf-8 -*-
###
# File: template.py
# Created Date: 2023-03-13 03:21:58
# Author: ysj
# Description:  python 模板方法
###

from abc import ABCMeta, abstractmethod


class ForgeSwordTemplate(metaclass=ABCMeta):
    score = 0

    def forge_sword(self):
        '''铸剑模板方法'''
        self.make_clay_model()
        self.dispense_materials()
        self.smelting_materials()
        self.water_shaping()
        self.repair_processing()

        self.result()

    @abstractmethod
    def make_clay_model(self):
        '''制做泥范, 每道工序20分'''
        print("制做泥范完美 +20分")
        self.score += 20

    @abstractmethod
    def dispense_materials(self):
        '''调剂材料, 每道工序20分'''
        print("调剂材料完美 +20分")
        self.score += 20

    @abstractmethod
    def smelting_materials(self):
        '''熔炼原料, 每道工序20分'''
        print("熔炼原料完美 +20分")
        self.score += 20

    @abstractmethod
    def water_shaping(self):
        '''浇灌成形, 每道工序20分'''
        print("浇灌成形完美 +20分")
        self.score += 20

    @abstractmethod
    def repair_processing(self):
        '''修治加工, 每道工序20分'''
        print("修治加工完美 +20分")
        self.score += 20

    def result(self):
        print("总分:", self.score)
        if self.score == 100:
            print("获得绝世好剑!!!")
        elif self.score > 80:
            print("获得一把好剑!!")
        else:
            print("获得一把村好剑!")

File: DesignPatterns/5-templateMethodPattern/golang/template.go
/*
 * File: template.go
 * Created Date: 2023-03-13 02:44:39
 * Author: ysj
 * Description:  golang 模板方法
 */
package main

import "fmt"

type ForgeSwordTemplate interface {
	ForgeSword()        // 铸剑模板方法
	makeClayModel()     // 制做泥范, 每道工序20分
	dispenseMaterials() // 调剂材料, 每道工序20分
	smeltingMaterials() // 熔炼原料, 每道工序20分
	waterShaping()      // 浇灌成形, 每道工序20分
	repairProcessing()  // 修治加工, 每道工序20分
}

type ForgeSwordBase struct {
	Score int
}

func (f *ForgeSwordBase) ForgeSword() {
	f.makeClayModel()
	f.dispenseMaterials()
	f.smeltingMaterials()
	f.waterShaping()
	f.repairProcessing()
	f.result()
}

func (f *ForgeSwordBase) makeClayModel() {
	fmt.Println("制做泥范完美 +20分")
	f.Score += 20
}

func (f *ForgeSwordBase) dispenseMaterials() {
	fmt.Println("调剂材料完美 +20分")
	f.Score += 20
}

func (f *ForgeSwordBase) smeltingMaterials() {
	fmt.Println("熔炼原料完美 +20分")
	f.Score += 20
}

func (f *ForgeSwordBase) waterShaping() {
	fmt.Println("浇灌成形完美 +20分")
	f.Score += 20
}

func (f *ForgeSwordBase) repairProcessing() {
	fmt.Println("修治加工完美 +20分")
	f.Score += 20
}

func (f *ForgeSwordBase) result() {
	fmt.Println("总分:", f.Score)
	if f.Score == 100 {
		fmt.Println("获得绝世好剑!!!")
	} else if f.Score > 80 {
		fmt.Println("获得一把好剑!!")
	} else {
		fmt.Println("获得一把村好剑!")
	}
}
